### PR TITLE
feat(design): add ability to theme the `@daffodil/design` library

### PR DESCRIPTION
**Pull request by damienwebdev**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
Currently, the colors and palettes of the design library are based on colors and styles that we've built out for the demo. 

#### Issue Numbers
Fixes: #68
Fixes: #113 
Fixes:  #151 
Touches On: #111 #114 #146 

## What is the new behavior?
This is an initial attempt at refactoring out those styles into a consistent pattern where we can swap colors and styles as we'd like for different daffodil implementations.

## Does this PR introduce a breaking change?
```
[x] Yes
[ ] No
```

I've moved the @section-divider and @cart-total mixins out of `@daffodil/design` into `@daffodil/demo`.

Additionally, I've added a few `scss` files under `daffio/src/scss` and `foundation-demo/src/scss` to help mediate the import layer between design and the two different projects. This is where you can add additional helper utilities and choose your theme colors/import them into components.

Dependent users will have to follow the same pattern, as I've removed the `$primary` `$secondary` and other related variables that were previously defined in `@daffodil/design` and used in `@daffodil/foundation-demo` / `@daffodil/daffio`.

## Other information

### CODE REVIEW

**Comment by damienwebdev** [comment]
@lderrickable @xelaint this is ready for review. I'm aware of a few minor edge cases that I have to resolve, but generally speaking, as this PR is large, I want to open it up for review early so that we can get some good feedback.

File: apps/demo/src/app/cart/components/cart-summary-wrapper/cart-summary-wrapper.component.scss [hunk 1]
```diff
@@ -1,4 +1,5 @@
-@import 'daff-util';
+@import 'utilities';
+@import "theme";
```
Comment (lderrickable) [inline apps/demo/src/app/cart/components/cart-summary-wrapper/cart-summary-wrapper.component.scss]: For some of these, I don't understand why `theme` is imported. Why is it imported here?

Comment (damienwebdev) [reply apps/demo/src/app/cart/components/cart-summary-wrapper/cart-summary-wrapper.component.scss]: I was copy/pasting to fix build errors. I just replaced daff-list-item with the both of them. It doesn't effect build times, but I can remove where un-needed.

Comment (lderrickable) [reply apps/demo/src/app/cart/components/cart-summary-wrapper/cart-summary-wrapper.component.scss]: Oh ok gotcha

**Comment by lderrickable** [comment]
It's pretty complicated on the inside, but seems pretty straightforward from the outside, which I like. I think it definitely needs docs though

File: apps/demo/src/scss/mixins/_cart-total.scss [hunk 2]
```diff
@@ -1,4 +1,5 @@
-@import "../typography/mixins";
```
Comment (xelaint) [inline apps/demo/src/scss/mixins/_cart-total.scss]: I think this mixin can be killed. It's not used anywhere and doesn't really make sense. We have a better implementation already in place using `@Input() emphasize: boolean;` in `cart-totals-item`.

File: apps/demo/src/scss/mixins/_dividers.scss [hunk 3]
```diff
@@ -1,3 +1,5 @@
+@import 'daff-theme';
```
Comment (xelaint) [inline apps/demo/src/scss/mixins/_dividers.scss]: I think this should be refactored into a `divider` component rather than a mixin so it can be used in broader instances.

**Comment by lderrickable** [comment]
Could you put a WIP on this if it's still a WIP?

**Comment by damienwebdev** [comment]
> Could you put a WIP on this if it's still a WIP?

I don't think this is still WIP, I'd like to merge as is. If there are changes that need to be made, please let me know if there's something that's confusing/you don't like/needs changed.

File: libs/design/src/scss/README.md [hunk 4]
```diff
@@ -1,2 +0,0 @@
-`daff-util.scss`
-Utility library that contains all the different mixins and functions.
```
Comment (lderrickable) [inline libs/design/src/scss/README.md]: needs docs

File: apps/demo/src/scss/mixins/_dividers.scss (hunk 3 above)
Comment (xelaint) [reply apps/demo/src/scss/mixins/_dividers.scss]: I agree we shouldn't do it in this PR, but there are many instances where I'd rather use a `divider` component than having to set border-radius on a specific container (that I might have override later).

Comment (lderrickable) [reply apps/demo/src/scss/mixins/_dividers.scss]: I thought the divider was the line that sometimes separates sections of the page. What does that have to do with border-radius or other components?

Comment (xelaint) [reply apps/demo/src/scss/mixins/_dividers.scss]: Oops, I mean border-top/border-bottom. I talked to Damien about this and it's not in the scope of this PR so I'm going to mark it as resolved.

File: libs/design/src/scss/typography/_mixins.scss [hunk 5]
```diff
@@ -1,11 +1,33 @@
...
+* @usage @include uppercase();
+*/
 @mixin uppercase() {
     text-transform: uppercase;
 }
 
+/**
+* @docs
+* 
+* Forces a line of test (with a fixed with) to ellipsis
```
Comment (xelaint) [inline libs/design/src/scss/typography/_mixins.scss]: Text  not test

File: apps/demo/src/scss/mixins/_cart-total.scss (hunk 2 above)
Comment (damienwebdev) [reply apps/demo/src/scss/mixins/_cart-total.scss]: @xelaint agreed removed in c36294e69967775496f33613cc1d556e30ff9446

File: apps/daffio/src/app/core/header/header-item/header-item.component.scss [hunk 6]
```diff
@@ -1,7 +1,7 @@
 @import 'daff-util';
 
 :host {
-  color: $black;
+  //color: $black;
```
[the remark this answers is not in the document]
Comment (damienwebdev) [reply apps/daffio/src/app/core/header/header-item/header-item.component.scss]: @lderrickable deleted.

File: apps/demo/src/app/cart/components/cart-summary-wrapper/cart-summary-wrapper.component.scss (hunk 1 above)
[the remark this answers is not in the document]
Comment (damienwebdev) [reply apps/demo/src/app/cart/components/cart-summary-wrapper/cart-summary-wrapper.component.scss]: Removed in 8b2579d92af1bda819a535674fb9efc826687fae

File: apps/demo/src/app/cart/components/cart-totals-item/cart-totals-item.component.scss [hunk 7]
```diff
@@ -1,4 +1,5 @@
-@import 'daff-util';
+@import 'utilities';
+@import 'theme';
```
[the remark this answers is not in the document]
Comment (damienwebdev) [reply apps/demo/src/app/cart/components/cart-totals-item/cart-totals-item.component.scss]: Removed in 8b2579d92af1bda819a535674fb9efc826687fae

File: apps/demo/src/app/cart/components/cart-totals/cart-totals.component.scss [hunk 8]
```diff
@@ -1,4 +1,5 @@
-@import 'daff-util';
+@import 'utilities';
+@import 'theme';
```
[the remark this answers is not in the document]
Comment (damienwebdev) [reply apps/demo/src/app/cart/components/cart-totals/cart-totals.component.scss]: Removed in 8b2579d92af1bda819a535674fb9efc826687fae

File: apps/demo/src/app/cart/components/cart/cart.component.scss [hunk 9]
```diff
@@ -1,4 +1,5 @@
-@import 'daff-util';
+@import 'utilities';
+@import "theme";
```
[the remark this answers is not in the document]
Comment (damienwebdev) [reply apps/demo/src/app/cart/components/cart/cart.component.scss]: Removed in 8b2579d92af1bda819a535674fb9efc826687fae

File: apps/demo/src/app/checkout/components/forms/address-form/components/address-form/address-form.component.scss [hunk 10]
```diff
@@ -1,4 +1,5 @@
-@import 'daff-util';
+@import 'utilities';
+@import 'theme';
```
[the remark this answers is not in the document]
Comment (damienwebdev) [reply apps/demo/src/app/checkout/components/forms/address-form/components/address-form/address-form.component.scss]: Removed in 8b2579d92af1bda819a535674fb9efc826687fae

File: apps/demo/src/app/checkout/components/payment/billing-summary/billing-summary.component.scss [hunk 11]
```diff
@@ -1,4 +1,5 @@
-@import 'daff-util';
+@import 'utilities';
+@import 'theme';
```
[the remark this answers is not in the document]
Comment (damienwebdev) [reply apps/demo/src/app/checkout/components/payment/billing-summary/billing-summary.component.scss]: Removed in 8b2579d92af1bda819a535674fb9efc826687fae

File: apps/demo/src/app/checkout/components/payment/payment-form/payment-form.component.scss [hunk 12]
```diff
@@ -1,4 +1,5 @@
-@import 'daff-util';
+@import 'utilities';
+@import 'theme';
```
[the remark this answers is not in the document]
Comment (damienwebdev) [reply apps/demo/src/app/checkout/components/payment/payment-form/payment-form.component.scss]: Removed in 8b2579d92af1bda819a535674fb9efc826687fae

File: libs/design/src/scss/README.md (hunk 4 above)
Comment (damienwebdev) [reply libs/design/src/scss/README.md]: Resolved in 6e0d8e5f09e11ac7c0fd9050ac33eb6a213be97d

File: libs/design/src/scss/typography/_mixins.scss (hunk 5 above)
Comment (damienwebdev) [reply libs/design/src/scss/typography/_mixins.scss]: Resolved in 6e0d8e5f09e11ac7c0fd9050ac33eb6a213be97d

File: apps/demo/src/app/checkout/components/shipping/shipping-options/components/shipping-options/shipping-options.component.scss [hunk 13]
```diff
@@ -1,5 +1,5 @@
-@import 'theme';
 @import 'utilities';
+@import 'theme';
```
Comment (xelaint) [inline apps/demo/src/app/checkout/components/shipping/shipping-options/components/shipping-options/shipping-options.component.scss]: Why did you add this back?

File: apps/daffio/src/app/content/homepage/component/homepage.component.scss [hunk 14]
```diff
@@ -1,11 +1,12 @@
-@import 'daff-util';
+@import 'theme';
```
Comment (xelaint) [inline apps/daffio/src/app/content/homepage/component/homepage.component.scss]: remove?

Comment (damienwebdev) [reply apps/daffio/src/app/content/homepage/component/homepage.component.scss]: Resolved in bee3c33bacc52d9bda5982546821edf17ab62b6d

File: apps/demo/src/app/checkout/components/shipping/shipping-options/components/shipping-options/shipping-options.component.scss (hunk 13 above)
Comment (damienwebdev) [reply apps/demo/src/app/checkout/components/shipping/shipping-options/components/shipping-options/shipping-options.component.scss]: It's necessary to use the $error color.

**Comment by damienwebdev** [comment]
I am also aware that this definitely needs a squash, so just in case, please do NOT merge until squashed.

**Comment by lderrickable** [comment]
I still feel like there should be more general documentation with how to use themeable. For someone just picking up `@daffodil/design`, I think it would be difficult to figure out how themeable should be used

**Comment by xelaint** [comment]
@lderrickable I think the README for theming can wait until `The Big Docs Update` milestone? I think at this point, we have enough docs to cover the basics. This PR is getting too large, especially if I'm making additional style updates or creating new components, I'll have to redo theming for them if this PR is still WIP.

**Comment by lderrickable** [comment]
Ok that's fine. Though I think that all the docs are written _except_ for the basics. In the future, especially for such a large feature, I'd really prefer that the basics were written so I can review the actual code better. It's helpful for me to understand what went through the coder's mind as he/she wrote it. Right now, it's pretty difficult to take a step back and understand how everything works

File: apps/daffio/src/scss/styles.scss [hunk 15]
```diff
@@ -4,6 +4,23 @@
...
+
+@import "theme";
+
+@include daff-theme($theme);
+
+@include daffio-homepage-theme($theme);
+@include daffio-pwa-theme($theme);
+@include daffio-simple-footer-theme($theme);
+@include daffio-sub-footer-theme($theme);
+@include daffio-feature-comparison-theme($theme);
```
Comment (lderrickable) [inline apps/daffio/src/scss/styles.scss]: I tried deleting lines 9-13 and 19-23 and the associated theming files, and it didn't seem to make a difference to the UI of daffio. What are they for?

Comment (xelaint) [reply apps/daffio/src/scss/styles.scss]: It removes the theming styles from these pages and components.

Comment (lderrickable) [reply apps/daffio/src/scss/styles.scss]: Including `daffio-pwa-theme($theme)` _removes_ theming styles from the pwa page? I have no idea what's going on

Comment (lderrickable) [reply apps/daffio/src/scss/styles.scss]: For reference "It" => "deleting the lines"

Comment (lderrickable) [inline apps/daffio/src/scss/styles.scss]: why are these imported in a global styling file? Wouldn't it be better to have them in each of their component scss files? Like the @include daffio-pwa-theme($theme); goes in the pwa.component.scss file

Comment (xelaint) [reply apps/daffio/src/scss/styles.scss]: I agree.

Comment (damienwebdev) [reply apps/daffio/src/scss/styles.scss]: @lderrickable the idea here is that these mixins include SCSS. We'd face two issues if we tried to include them in the component file rather than at the global level:

1. Specificity, components that are ViewEncapsulated get `[nghost]` on them which would
2. Ability to do multi-theme without having to specify classes all over the place, we can quickly centralize them at the top of the app and apply a single class globally.

**Comment by lderrickable** [comment]
lgtm

**Comment by damienwebdev** [comment]
@lderrickable @xelaint squashed, needs approval.
